Support registering inert attributes and attribute tools using crate-level attributes

petrochenkov opened the pull request:
And remove `#[feature(custom_attribute)]`.
(`rustc_plugin::Registry::register_attribute` is not removed yet, I'll do it in a follow up PR.)

```rust
#![register_attr(my_attr)]
#![register_tool(my_tool)]

#[my_attr] // OK
#[my_tool::anything] // OK
fn main() {}
```

---
Some tools (`rustfmt` and `clippy`) used in tool attributes are hardcoded in the compiler.
We need some way to introduce them without hardcoding as well.

This PR introduces a way to do it with a crate level attribute.
The previous attempt to introduce them through command line (https://github.com/rust-lang/rust/pull/57921) met some resistance.

This probably needs to go through an RFC before stabilization.
However, I'd prefer to land *this* PR without an RFC to able to remove `#[feature(custom_attribute)]` and `Registry::register_attribute` while also providing a replacement.

---
`register_attr` is a direct replacement for `#![feature(custom_attribute)]` (https://github.com/rust-lang/rust/issues/29642), except it doesn't rely on implicit fallback from unresolved attributes to custom attributes (which was always hacky and is the primary reason for the removal of `custom_attribute`) and requires registering the attribute explicitly.
It's not clear whether it should go through stabilization or not.
It's quite possible that all the uses should migrate to `#![register_tool]` (https://github.com/rust-lang/rust/issues/66079) instead.

---

Details:
- The naming is `register_attr`/`register_tool` rather than some `register_attributes` (plural, no abbreviation) for consistency with already existing attributes like `cfg_attr`, or `feature`, etc.
---
Previous attempt: https://github.com/rust-lang/rust/pull/57921
cc https://github.com/rust-lang/rust/issues/44690
Tracking issues: #66079 (`register_tool`), #66080 (`register_attr`)
Closes https://github.com/rust-lang/rust/issues/29642

petrochenkov committed:
Support registering attributes and attribute tools using crate-level attributes

petrochenkov committed:
Remove `#[feature(custom_attribute)]`

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

Centril commented on src/librustc_resolve/macros.rs:
```diff
@@ -94,6 +95,45 @@ fn fast_print_path(path: &ast::Path) -> Symbol {
     }
 }
 
+fn registered_idents(
```
A doc comment would be great. :)

Centril commented on src/librustc_resolve/macros.rs:
```diff
@@ -558,12 +607,14 @@ impl<'a> Resolver<'a> {
+                        let binding = (Res::ToolMod,
+                                       ty::Visibility::Public, ident.span, ExpnId::root())
+                                       .to_name_binding(this.arenas);
+                        Ok((binding, Flags::PRELUDE))
```
Some of this logic is repeated above and in other places; could it be de-duplicated?

Centril commented on src/librustc_resolve/lib.rs:
```diff
@@ -1471,6 +1477,7 @@ impl<'a> Resolver<'a> {
                 Scope::MacroRules(..) => true,
                 Scope::CrateRoot => true,
                 Scope::Module(..) => true,
+                Scope::RegisteredAttrs => true,
```
Or-patterns in various arms would be good here.

petrochenkov replied:
I thinks it's better to list scopes in the same order here than merge arms.

Centril commented on src/libsyntax/feature_gate/active.rs:
```diff
@@ -533,6 +530,12 @@ declare_features! (
+    (active, register_attr, "1.41.0", Some(29642), None),
+
+    /// Allows using the `#[register_attr]` attribute.
+    (active, register_tool, "1.41.0", Some(44690), None),
```
I would prefer new fresh tracking issues for both of these to have more targeted discussion and triage.

petrochenkov replied:
Tracking issue for tools - https://github.com/rust-lang/rust/issues/66079.

petrochenkov replied:
Tracking issue for attrs - https://github.com/rust-lang/rust/issues/66080.

petrochenkov committed:
resolve: Remove some bits relevant only to legacy plugins

They are unstable and going to be removed anyway and the removed code would complicate the next commit

petrochenkov committed:
resolve: Factor out some common binding creation functionality

petrochenkov committed:
Address review comments

fbstj commented on src/test/ui/attributes/register-attr-tool-fail.rs:
```diff
@@ -0,0 +1,13 @@
+
+#![register_attr(attr, attr)] //~ ERROR attribute `attr` was already registered
+#![register_tool(tool, tool)] //~ ERROR tool `tool` was already registered
+
```
what's the result of attr vs tool collision?

```rust
#![register_attr(something)]
#![register_tool(something)]
```

petrochenkov replied:
They do not collide because the attribute is in macro namespace and the tool is in type namespace.

fbstj commented on src/test/ui/attributes/register-attr-tool.rs:
```diff
@@ -0,0 +1,19 @@
+#![cfg_attr(foo, register_tool(conditional_tool))]
+
+#[attr]
+#[tool::attr]
```
what if there's just `#[tool]` ?

petrochenkov replied:
```rust
#[rustfmt] // error: cannot find attribute `rustfmt` in this scope
fn main() {}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a34779363ede9113bf4f8574f69f42da

petrochenkov replied:
We already have a test case for this in `ui/tool-attributes/tool-attributes-misplaced-1.rs`.

matthewjasper reviewed (commented):
This mostly looks good. There are a few things that I would like to see tests for:

* `#[no_implicit_prelude]` should hide these attributes.
* Using the `register*` attributes on something other than a crate should be a warning/error.
* Whether registed attributes can be imported with `use`.

petrochenkov commented on src/libsyntax/feature_gate/builtin_attrs.rs:
```diff
@@ -329,6 +329,14 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
     gated!(ffi_returns_twice, Whitelisted, template!(Word), experimental!(ffi_returns_twice)),
     gated!(track_caller, Whitelisted, template!(Word), experimental!(track_caller)),
+    gated!(
+        register_attr, Whitelisted, template!(List: "attr1, attr2, ..."),
```
Good catch, I copypasted the item and forgot to change the attribute type.

JohnCSimon commented:
Ping from triage - 
@petrochenkov - this PR hasn't been updated in a few days
cc: @matthewjasper 
thanks!

petrochenkov commented:
Updated with the comments addressed.

matthewjasper commented:
@bors r+

bors commented:
:pushpin: Commit 83f553c95c4a7cfd44ec00ce056708a7022b7bf7 has been approved by `matthewjasper`

<!-- @bors r=matthewjasper 83f553c95c4a7cfd44ec00ce056708a7022b7bf7 -->
<!-- homu: {"type":"Approved","sha":"83f553c95c4a7cfd44ec00ce056708a7022b7bf7","approver":"matthewjasper"} -->

bors commented:
:hourglass: Testing commit 83f553c95c4a7cfd44ec00ce056708a7022b7bf7 with merge b3f59a942069b71a14d6eb5991f582c054864de2...
<!-- homu: {"type":"BuildStarted","head_sha":"83f553c95c4a7cfd44ec00ce056708a7022b7bf7","merge_sha":"b3f59a942069b71a14d6eb5991f582c054864de2"} -->

rust-highfive commented:
The job `x86_64-msvc-cargo` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13149&view=logs&jobId=34a2f069-38a6-567f-c087-79a147504c7a) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13149/logs/1536), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13149/logs/1536)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-10T05:34:52.6300872Z 
2019-11-10T05:34:52.6300903Z 
2019-11-10T05:34:52.6300950Z 
2019-11-10T05:34:52.6300982Z 
2019-11-10T05:34:52.6301056Z #![register_attr(my_attr)]
2019-11-10T05:34:52.6301127Z #![register_tool(my_tool)]
2019-11-10T05:34:52.6301211Z #[my_attr] // OK
2019-11-10T05:34:52.6301278Z #[my_tool::anything] // OK
2019-11-10T05:34:52.6301389Z (`rustc_plugin::Registry::register_attribute` is not removed yet, I'll do it in a follow up PR.)
2019-11-10T05:34:52.6301543Z - The naming is `register_attr`/`register_tool` rather than some `register_attributes` (plural, no abbreviation) for consistency with already existing attributes like `cfg_attr`, or `feature`, etc.
2019-11-10T05:34:52.6301753Z ---
2019-11-10T05:34:52.6301807Z ---
2019-11-10T05:34:52.6301874Z ---
2019-11-10T05:34:52.6301932Z AGENT_BUILDDIRECTORY=D:\a\1
---
2019-11-10T05:34:52.6304252Z ANT_HOME=C:\ProgramData\chocolatey\lib\ant\apache-ant-1.10.5
2019-11-10T05:34:52.6304329Z APPDATA=C:\Users\VssAdministrator\AppData\Roaming
2019-11-10T05:34:52.6304424Z AZURE_EXTENSION_DIR=C:\Program Files\Common Files\AzureCliExtensionDirectory
2019-11-10T05:34:52.6305400Z AZURE_HTTP_USER_AGENT=VSTS_d439fc94-e01f-4249-b63e-d8392bc0247c_build_10_0
2019-11-10T05:34:52.6305559Z And remove `#[feature(custom_attribute)]`.
2019-11-10T05:34:52.6305723Z BOOST_ROOT_1_69_0=C:\Program Files\Boost\1.69.0
2019-11-10T05:34:52.6305788Z BUILD_ARTIFACTSTAGINGDIRECTORY=D:\a\1\a
2019-11-10T05:34:52.6305867Z BUILD_BINARIESDIRECTORY=D:\a\1\b
2019-11-10T05:34:52.6305943Z BUILD_BUILDID=13149
---
2019-11-10T05:34:52.6310810Z GeckoWebDriver=C:\SeleniumWebDrivers\GeckoDriver
2019-11-10T05:34:52.6310901Z HOME=/c/Users/VssAdministrator
2019-11-10T05:34:52.6310987Z HOMEDRIVE=C:
2019-11-10T05:34:52.6311050Z HOMEPATH=\Users\VssAdministrator
2019-11-10T05:34:52.6311574Z However, I'd prefer to land *this* PR without an RFC to able to remove `#[feature(custom_attribute)]` and `Registry::register_attribute` while also providing a replacement.
2019-11-10T05:34:52.6311817Z INPUT_ARGUMENTS=
2019-11-10T05:34:52.6311881Z ImageVersion=20191028.1
2019-11-10T05:34:52.6311881Z ImageVersion=20191028.1
2019-11-10T05:34:52.6311981Z It's not clear whether it should go through stabilization or not.
2019-11-10T05:34:52.6312115Z It's quite possible that all the uses should migrate to `#![register_tool]` (https://github.com/rust-lang/rust/issues/66079) instead.
2019-11-10T05:34:52.6312346Z JAVA_HOME_11_X64=C:\Program Files\Java\zulu-11-azure-jdk_11.33.15-11.0.4-win_x64
2019-11-10T05:34:52.6312448Z JAVA_HOME_7_X64=C:\Program Files\Java\zulu-7-azure-jdk_7.31.0.5-7.0.232-win_x64
2019-11-10T05:34:52.6312567Z JAVA_HOME_8_X64=C:\Program Files\Java\zulu-8-azure-jdk_8.40.0.25-8.0.222-win_x64
2019-11-10T05:34:52.6312669Z LOCALAPPDATA=C:\Users\VssAdministrator\AppData\Local
---
2019-11-10T05:34:52.6321560Z SYSTEM_TEAMPROJECTID=e71b0ddf-dd27-435a-873c-e30f86eea377
2019-11-10T05:34:52.6321663Z SYSTEM_TIMELINEID=e35078b8-1d30-49c3-8229-28167d9adf69
2019-11-10T05:34:52.6321737Z SYSTEM_TOTALJOBSINPHASE=16
2019-11-10T05:34:52.6321822Z SYSTEM_WORKFOLDER=D:\a
2019-11-10T05:34:52.6321929Z Some tools (`rustfmt` and `clippy`) used in tool attributes are hardcoded in the compiler.
2019-11-10T05:34:52.6322036Z Support registering inert attributes and attribute tools using crate-level attributes
2019-11-10T05:34:52.6322211Z TEMP=/tmp
2019-11-10T05:34:52.6322283Z TERM=cygwin
2019-11-10T05:34:52.6322343Z TF_BUILD=True
2019-11-10T05:34:52.6322425Z TMP=/tmp
2019-11-10T05:34:52.6322425Z TMP=/tmp
2019-11-10T05:34:52.6322503Z TOOLSTATE_ISSUES_API_URL=https://api.github.com/repos/rust-lang/rust/issues
2019-11-10T05:34:52.6322598Z TOOLSTATE_PUBLISH=1
2019-11-10T05:34:52.6322703Z TOOLSTATE_REPO=https://github.com/rust-lang-nursery/rust-toolstate
2019-11-10T05:34:52.6322818Z The previous attempt to introduce them through command line (https://github.com/rust-lang/rust/pull/57921) met some resistance.
2019-11-10T05:34:52.6322945Z This PR introduces a way to do it with a crate level attribute.
2019-11-10T05:34:52.6323155Z This probably needs to go through an RFC before stabilization.
2019-11-10T05:34:52.6323252Z Tracking issues: #66079 (`register_tool`), #66080 (`register_attr`)
2019-11-10T05:34:52.6323403Z USERDOMAIN=fv-az363
2019-11-10T05:34:52.6323475Z USERDOMAIN_ROAMINGPROFILE=fv-az363
2019-11-10T05:34:52.6323536Z USERNAME=VssAdministrator
2019-11-10T05:34:52.6323616Z USERPROFILE=C:\Users\VssAdministrator
2019-11-10T05:34:52.6323616Z USERPROFILE=C:\Users\VssAdministrator
2019-11-10T05:34:52.6323679Z VCPKG_INSTALLATION_ROOT=C:\vcpkg
2019-11-10T05:34:52.6323753Z VCVARS_BAT=vcvars64.bat
2019-11-10T05:34:52.6323830Z VS140COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\
2019-11-10T05:34:52.6323924Z VSTS_AGENT_PERFLOG=c:\vsts\perflog
2019-11-10T05:34:52.6324012Z VSTS_PROCESS_LOOKUP_ID=vsts_a793f582-7718-41fe-a734-0b3bb3114a89
2019-11-10T05:34:52.6324079Z WINDIR=C:\windows
2019-11-10T05:34:52.6324156Z WIX=C:\Program Files (x86)\WiX Toolset v3.11\
2019-11-10T05:34:52.6324231Z We need some way to introduce them without hardcoding as well.
2019-11-10T05:34:52.6324367Z ```
2019-11-10T05:34:52.6324429Z ```rust
2019-11-10T05:34:52.6324429Z ```rust
2019-11-10T05:34:52.6324587Z `register_attr` is a direct replacement for `#![feature(custom_attribute)]` (https://github.com/rust-lang/rust/issues/29642), except it doesn't rely on implicit fallback from unresolved attributes to custom attributes (which was always hacky and is the primary reason for the removal of `custom_attribute`) and requires registering the attribute explicitly.
2019-11-10T05:34:52.6324827Z cc https://github.com/rust-lang/rust/issues/44690
2019-11-10T05:34:52.6324960Z fn main() {}
2019-11-10T05:34:52.6328746Z 
2019-11-10T05:34:52.6329094Z disk usage:
---
2019-11-10T07:45:33.4077821Z [RUSTC-TIMING] lazycell test:false 0.139
2019-11-10T07:45:33.4103237Z    Compiling hex v0.4.0
2019-11-10T07:45:36.8057313Z [RUSTC-TIMING] hex test:false 3.386
2019-11-10T07:45:36.8155967Z    Compiling glob v0.3.0
2019-11-10T07:45:37.4832023Z memory allocation of 2147483656 bytes failed[RUSTC-TIMING] hex test:false 4.275
2019-11-10T07:45:37.4943635Z error: could not compile `hex`.
2019-11-10T07:45:38.7754599Z [RUSTC-TIMING] glob test:false 1.953
2019-11-10T07:45:38.8158391Z error: build failed
2019-11-10T07:45:38.8225517Z command did not execute successfully: "D:\\a\\1\\s\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "build" "-Zconfig-profile" "--target" "x86_64-pc-windows-msvc" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--locked" "--color" "always" "--manifest-path" "D:\\a\\1\\s\\src/tools/cargo\\Cargo.toml" "--features" "rustc-workspace-hack/all-static" "--message-format" "json-render-diagnostics"
2019-11-10T07:45:38.8225922Z expected success, got: exit code: 101
2019-11-10T07:45:38.8225922Z expected success, got: exit code: 101
2019-11-10T07:45:38.9372893Z failed to run: D:\a\1\s\build\bootstrap\debug\bootstrap test src/tools/cargotest src/tools/cargo
2019-11-10T07:45:38.9373036Z Build completed unsuccessfully in 1:59:28
2019-11-10T07:45:39.0125968Z == clock drift check ==
2019-11-10T07:45:39.1137962Z   local time: Sun Nov 10 07:45:39 CUT 2019
2019-11-10T07:45:39.5337009Z   network time: Sun, 10 Nov 2019 07:45:39 GMT
2019-11-10T07:45:39.5355087Z == end clock drift check ==
2019-11-10T07:45:39.6089271Z 
2019-11-10T07:45:40.0425355Z ##[error]Bash exited with code '1'.
2019-11-10T07:45:40.1031801Z ##[section]Starting: Checkout
2019-11-10T07:45:40.1902148Z ==============================================================================
2019-11-10T07:45:40.1902295Z Task         : Get sources
2019-11-10T07:45:40.1902380Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13149)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13149","builder_name":"checks-azure"} -->

petrochenkov commented:
```
2019-11-10T07:45:37.4832023Z memory allocation of 2147483656 bytes failed[RUSTC-TIMING] hex test:false 4.275
2019-11-10T07:45:37.4943635Z error: could not compile `hex`.
```
Wat.
@bors retry

bors commented:
:hourglass: Testing commit 83f553c95c4a7cfd44ec00ce056708a7022b7bf7 with merge 3fc30d884ae0c988d98452a06737705cfe34806a...
<!-- homu: {"type":"BuildStarted","head_sha":"83f553c95c4a7cfd44ec00ce056708a7022b7bf7","merge_sha":"3fc30d884ae0c988d98452a06737705cfe34806a"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13190)
Approved by: matthewjasper
Pushing 3fc30d884ae0c988d98452a06737705cfe34806a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"matthewjasper","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13190"},"merge_sha":"3fc30d884ae0c988d98452a06737705cfe34806a"} -->
